src/validators/validators.ts

```diff
@@ -0,0 +1,11 @@
+// Validators are functions that return `undefined` or `false`
+// when they pass, and return a string with an explanation when
+// they fail.
+
+export type Validator = (value: any) => string | void | false;
+
+export const isString: Validator = val =>
+  typeof val !== "string" && "must be a string";
```

Comment:
While the validator code is a lot simpler this way, and we've gotten rid of the "validator" function dependency, I think this makes the API significantly less intuitive. I'm not sure whether or not it's worth it. What do you think, @joeldenning ?

Reply:
Yeah it's a bit less intuitive. Maybe make the return value more explicit?

```js
if (isValid) {
  return {valid: true}
} else {
  return {valid: false, reason: 'must start with "thi"'}
}
```

Reply:
I think that's a much clunkier API than
```
validator(isValid, "it's not valid!")
```

I'm starting to think that the intuitiveness of the API with the wrapping `validator` function justifies its existence.

Though another possible API would be that validators should be functions that return an Error when validation fails:
```
(val) => val < 10 || new Error("must be less than ten")
```

Not sure if this is an anti-pattern, but at least it's concise, easy to understand, and dependency-free.

Reply:
Well, I guess it could also be
```
(val) => val < 10 || "must be less than ten"
```

where any string return value signifies a failure. Is that better? Is it worse?

Reply:
> I think that's a much clunkier API

Having to call `validator()` dozens of times in dozens of repos is a lot clunkier than returning `{valid: true}` one time inside of the `isString` validator.

What I didn't like about the `validator()` function is that (if I understood it correctly) it has to be used by everyone every time in every repo. It couples the schema definition to the version of esm-module-config that is being used. You cannot test a schema without mocking the `validator` function

```js
:sad:
import { validator, isString } from '@openmrs/esm-module-config'

defineConfigSchema({
  thing: validator(val => typeof val === 'string'),
  otherThing: validator(isString),
  another: validator(isNumber)
})
```

```js
:happy:
defineConfigSchema({
  thing: val => {valid: typeof val === 'string', reason: "Need a string"},
  otherThing: isString,
  another: isNumber
})
```

Reply:
To clarify, the original API involving the "validator" function would only require it when defining your own validator (which should, to be fair, happen often):
```
import { validator, isString, isNumber } from '@openmrs/esm-module-config'

defineConfigSchema({
  thing: { default: 0, validators: [validator(val => val < 10, "must be less than ten")]},
  otherThing: { default: "hey", validators: [isString]},
  another: { default: 1, validators: [isNumber]}
})
```

Because the pre-packaged validators come pre-wrapped in `validator`.

My alternative proposals, above, would yield usage like
```
import { isString, isNumber } from '@openmrs/esm-module-config'

defineConfigSchema({
  thing: { default: 0, validators: [val => val < 10 || new Error("must be less than ten")]},
  otherThing: { default: "hey", validators: [isString]},
  another: { default: 1, validators: [isNumber]}
})
```

or alternatatively

```
import { isString, isNumber } from '@openmrs/esm-module-config'

defineConfigSchema({
  thing: { default: 0, validators: [val => val < 10 || "must be less than ten"]},
  otherThing: { default: "hey", validators: [isString]},
  another: { default: 1, validators: [isNumber]}
})
```

Your API with `valid` and `reason` is also nice. I think I would be happiest with either that or the `new Error` API. Now that it's clear what we're comparing, what do you think?

Reply:
let's go with your original api. Seems easiest.

Reply:
Ok, I've re-implemented that API, but I've simplified the implementation along the lines of your recommendation. This is ready for re-review.